path_planning.py

```diff
@@ -0,0 +1,321 @@
+    
+   
+    node = goal
+    bestcost,point = best_costs.get(node)   
+    retpath: List[Point] = []
+    while(1):       
+        cost,point = best_costs.get(node)          
+        retpath.insert(0,node)
+        if(compare_points(point,start)) :
+            retpath.insert(0,point)
+            break;
+        node = point
+     
+    print("bestcost" ,bestcost    )
+    print("path" ,retpath    )
+   
+ 
+    return (visit_count, bestcost, retpath)    
+
+def best_first(terrain:Map, start:Point, goal:Point) -> Tuple[int, int, Optional[List[Point]]]:
+    # In the open list we use heuristic values as the priority
+    open_list: List[Tuple[int, Point]] = [(manhattan_distance(start, goal), start)]
+    # But in best_costs we still want to track real costs
+    best_costs: Dict[Point, Tuple[int, Point]] = {start:(0, start)}
+    visit_count = 0
+    found = False
+    while open_list:
+        # Dijkstra's search uses the priority queue data structure
+        dummycost, node = heapq.heappop(open_list)              
+        #print(node)
+       
+        if(compare_points(node,goal)):            
+            #print(node,goal)
+            found = True;
+            break;
+        visit_count += 1        
+        
+        neighbors = find_neighbors(terrain, node)       
+        for neighbor, neighbor_cost in neighbors:
+            if(neighbor in  best_costs):
+                #print("found=" , neighbor)     
+                continue
+            else:    
+                    cost,point = best_costs.get(node)#print(neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )                   
+                    heapq.heappush (open_list,  ( manhattan_distance(neighbor, goal),  neighbor ) )     
+                    
+    if(found == False) :                   
+        return (visit_count, -1, None)
+    
+
+   
+    node = goal
+    bestcost,point = best_costs.get(node)   
+    retpath: List[Point] = []
+    while(1):       
+        cost,point = best_costs.get(node)          
+        retpath.insert(0,node)
+        if(compare_points(point,start)) :
+            retpath.insert(0,point)
+            break;
+        node = point
+     
+    print("bestcost" ,bestcost    )
+    print("path" ,retpath    )
+   
+ 
+    return (visit_count, bestcost, retpath)    
+
+ 
+
+def dijkstra(terrain:Map, start:Point, goal:Point) -> Tuple[int, int, Optional[List[Point]]]:
+    open_list: List[Tuple[int, Point]] = [(0, start)]         
+    best_costs: Dict[Point, Tuple[int, Point]] = {start:(0, start)}
+    visit_count = 0
+    found = False
+    while open_list:
+        # Dijkstra's search uses the priority queue data structure
+        cost, node = heapq.heappop(open_list)       
+        #print(node)
+       
+        if(compare_points(node,goal)):
+            #print ("found it") 
+            found = True;
+            break;
+                
+        visit_count += 1

```

Comment:
Same as BFS - does this undercount by 1 because it increments after goal checking?